File: examples/go/basic-bot.go
package main

import (
  "fmt"
  "os"
  "os/signal"
  "strings"
  "syscall"

  "github.com/bwmarrin/discordgo"
  "github.com/joho/godotenv"
)

func main() {
  godotenv.Load()
  token := os.Getenv("TOKEN")
  prefix := os.Getenv("PREFIX")
  if prefix == "" {
    prefix = "!"
  }

  dg, err := discordgo.New("Bot " + token)
  if err != nil {
    fmt.Println("Error creating session:", err)
    return
  }

  dg.AddHandler(func(s *discordgo.Session, r *discordgo.Ready) {
    fmt.Println("Bot is ready as", s.State.User.Username)
    s.AddHandler(func(s *discordgo.Session, i *discordgo.InteractionCreate) {
      if i.ApplicationCommandData().Name == "ping" {
        s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
          Type: discordgo.InteractionResponseChannelMessageWithSource,
          Data: &discordgo.InteractionResponseData{Content: "Pong!"},
        })
      }
    })
  })

  dg.AddHandler(func(s *discordgo.Session, m *discordgo.MessageCreate) {
    if m.Author.Bot { return }
    if !strings.HasPrefix(m.Content, prefix) { return }

    args := strings.Fields(strings.TrimPrefix(m.Content, prefix))
    if len(args) == 0 { return }

    switch strings.ToLower(args[0]) {
    case "ping":
      s.ChannelMessageSend(m.ChannelID, "Pong!")
    }
  })

  dg.Identify.Intents = discordgo.IntentsAllWithoutPrivileged | discordgo.Intents.MessageContent

  err = dg.Open()
  if err != nil {
    fmt.Println("Error opening connection:", err)
    return
  }

  fmt.Println("Bot is running. Press CTRL+C to exit.")
  sc := make(chan os.Signal, 1)
  signal.Notify(sc, syscall.SIGINT, syscall.SIGTERM)
  <-sc
  dg.Close()
}

File: examples/rust/basic-bot.rs
use serenity::all::*;
use serenity::prelude::*;
use std::env;

struct Handler;

#[serenity::async_trait]
impl EventHandler for Handler {
  async fn ready(&self, ctx: Context, ready: Ready) {
    println!("Logged in as {}", ready.user.name);

    let command = CreateCommand::new("ping").description("Replies with Pong!");
    Command::create_global_command(&ctx.http, command).await.unwrap();
  }

  async fn interaction_create(&self, ctx: Context, interaction: Interaction) {
    if let Some(command) = interaction.as_command() {
      if command.data.name == "ping" {
        command.create_response(&ctx.http, CreateInteractionResponse::Message(
          CreateInteractionResponseMessage::new().content("Pong!"),
        )).await.unwrap();
      }
    }
  }

  async fn message(&self, ctx: Context, msg: Message) {
    if msg.author.bot { return; }

    let prefix = env::var("PREFIX").unwrap_or_else(|_| "!".to_string());
    if !msg.content.starts_with(&prefix) { return; }

    let args: Vec<&str> = msg.content[prefix.len()..].split_whitespace().collect();
    if args.is_empty() { return; }

    match args[0] {
      "ping" => { msg.channel_id.say(&ctx.http, "Pong!").await.unwrap(); }
      _ => {}
    }
  }
}

#[tokio::main]
async fn main() {
  dotenv::dotenv().ok();
  let token = env::var("TOKEN").expect("Expected TOKEN in environment");

  let intents = GatewayIntents::all();
  let mut client = Client::builder(&token, intents)
    .event_handler(Handler)
    .await
    .expect("Error creating client");

  if let Err(e) = client.start().await {
    println!("Client error: {:?}", e);
  }
}
